dbms/src/Functions/FunctionsStringHash.cpp

```diff
@@ -0,0 +1,585 @@
+#include <Functions/FunctionsStringHash.h>
+
+#include <Functions/ExtractString.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionsHashing.h>
+#include <Common/HashTable/ClearableHashMap.h>
+#include <Common/HashTable/Hash.h>
+
+#include <Core/Defines.h>
+
+#include <bitset>
+#include <tuple>
+#include <common/unaligned.h>
+
+namespace DB
+{
+struct Hash
+{
+    static ALWAYS_INLINE inline UInt64 ngramASCIIHash(const UInt8 * code_points)
+    {
+        return intHashCRC32(unalignedLoad<UInt32>(code_points));
+    }
+
+    static ALWAYS_INLINE inline UInt64 ngramUTF8Hash(const UInt32 * code_points)
+    {
+        UInt64 combined = (static_cast<UInt64>(code_points[0]) << 32) | code_points[1];
+#ifdef __SSE4_2__
+        return _mm_crc32_u64(code_points[2], combined);
+#else
+        return (intHashCRC32(combined) ^ intHashCRC32(code_points[2]));
+#endif
+    }
+
+    static ALWAYS_INLINE inline UInt64 wordShinglesHash(const UInt64 * hashes, const size_t & size, const size_t & offset)
+    {
+        UInt64 res = 0;
+        UInt8 flag = 0;
+        for (size_t i = offset; i < size; ++i)
+        {
+            if (flag)
+                res &= intHashCRC32(hashes[i]);
```

Comment:
Not a good hash function. Better to calculate CRC32 (_mm_crc32_u64) by chain (starting with -1 as initial value).
Look at `UInt256HashCRC32` as an example.